### PR TITLE
[3006.x] Fix pkg.install when using a port in the url.

### PR DESCRIPTION
**Pull request by Ch3LL**
### What does this PR do?
Removes `:` from url when creating the cache directory to save the file. `apt` does not handle `:` in directory names and windows does not allow it.

### What issues does this PR fix or reference?
Fixes: https://github.com/saltstack/salt/issues/64516

### Previous Behavior
Could not install a package with a source url with an port in it.

### New Behavior
Can now install a package with a source with a url and a port in it.

### CODE REVIEW

File: tests/pytests/unit/fileclient/test_fileclient.py [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+                    assert c_ref_itr == os.sep + os.sep.join(
+                        ["__test__", "files", "base", "testfile"]
+                    )
+
+
+def test_cache_raises_exception_on_non_eexist_ioerror(client_opts):
+    """
+    If makedirs raises other than EEXIST errno, an exception should be raised.
+    """
+    with patch("os.path.isfile", lambda prm: False):
```
Comment (s0undt3ch) [inline tests/pytests/unit/fileclient/test_fileclient.py:159]:
```suggestion
    with patch("os.path.isfile", return_value=False):
```
?

File: tests/pytests/unit/fileclient/test_fileclient.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+@pytest.fixture
+def configure_loader_modules(tmp_path, mocked_opts):
+    return {fileclient: {"__opts__": mocked_opts}}
+
+
+@pytest.fixture
+def file_client(mocked_opts):
+    client = fileclient.Client(mocked_opts)
+    yield client
+    del client
```
Comment (s0undt3ch) [inline tests/pytests/unit/fileclient/test_fileclient.py:40]:
```suggestion
def file_client(mocked_opts):
    client = fileclient.Client(mocked_opts)
    try:
        yield client
    finally:
        del client
```

File: tests/pytests/unit/fileclient/test_fileclient.py [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+    async def mock_auth():
+        return True
+
+    def mock_dumps(*args):
+        return b"meh"
+
+    with fileclient.get_file_client(minion_opts) as client:
+        # Authenticate must return true
+        client.auth.authenticate = mock_auth
+        # Crypticla must return bytes to pass to transport.RequestClient.send
```
Comment (s0undt3ch) [inline tests/pytests/unit/fileclient/test_fileclient.py:135]:
```suggestion
        # Crypticle must return bytes to pass to transport.RequestClient.send
```

File: tests/pytests/unit/fileclient/test_fileclient.py [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+        client.auth._crypticle.dumps = mock_dumps
+        with pytest.raises(salt.exceptions.SaltClientError):
+            client.file_list()
+
+
+def test_cache_skips_makedirs_on_race_condition(client_opts):
+    """
+    If cache contains already a directory, do not raise an exception.
+    """
+    with patch("os.path.isfile", lambda prm: False):
```
Comment (s0undt3ch) [inline tests/pytests/unit/fileclient/test_fileclient.py:146]:
```suggestion
    with patch("os.path.isfile", return_value=False):
```
?

File: tests/pytests/unit/fileclient/test_fileclient_cache.py [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+    # Create the CACHE_ROOT
+    _new_dir(cache_root)
+
+
+def test_cache_dir(mocked_opts, minion_opts):
+    """
+    Ensure entire directory is cached to correct location
+    """
+    patched_opts = {x: y for x, y in minion_opts.items()}
+    patched_opts.update(mocked_opts)
```
Comment (s0undt3ch) [inline tests/pytests/unit/fileclient/test_fileclient_cache.py:109]:
```suggestion
    patched_opts = minion_opts.copy()
    patched_opts.update(mocked_opts)
```
In fact, the `minion_opts` fixture is generated(or copied) for every test function, so, you might not even need to copy it...

File: tests/pytests/unit/fileclient/test_fileclient_cache.py [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+
+def test_cache_dir_with_alternate_cachedir_and_absolute_path(
+    mocked_opts, minion_opts, tmp_path
+):
+    """
+    Ensure entire directory is cached to correct location when an alternate
+    cachedir is specified and that cachedir is an absolute path
+    """
+    patched_opts = {x: y for x, y in minion_opts.items()}
+    patched_opts.update(mocked_opts)
```
Comment (s0undt3ch) [inline tests/pytests/unit/fileclient/test_fileclient_cache.py:147]:
```suggestion
    patched_opts = {x: y for x, y in minion_opts.items()}
    patched_opts.update(mocked_opts)
```
Same as above

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

**Review by garethgreenaway — DISMISSED** [review]

File: tests/pytests/unit/fileclient/test_fileclient_cache.py [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+                assert subdir_file in content
+                assert SUBDIR in content
+                assert saltenv in content
+
+
+def test_cache_file(mocked_opts, minion_opts):
+    """
+    Ensure file is cached to correct location
+    """
+    patched_opts = {x: y for x, y in minion_opts.items()}
```
Comment (s0undt3ch) [inline tests/pytests/unit/fileclient/test_fileclient_cache.py:220]:
```suggestion
    patched_opts = minion_opts.copy()
```
Sorry didn't want to sound repetitive on my previous review.

File: tests/pytests/unit/fileclient/test_fileclient_cache.py [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+            log.debug("content = %s", content)
+            assert saltenv in content
+
+
+def test_cache_file_with_alternate_cachedir_and_relative_path(mocked_opts, minion_opts):
+    """
+    Ensure file is cached to correct location when an alternate cachedir is
+    specified and that cachedir is a relative path
+    """
+    patched_opts = {x: y for x, y in minion_opts.items()}
```
Comment (s0undt3ch) [inline tests/pytests/unit/fileclient/test_fileclient_cache.py:279]: Same as above

File: tests/pytests/unit/fileclient/test_fileclient_cache.py [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+            log.debug("cache_loc = %s", cache_loc)
+            log.debug("content = %s", content)
+            assert saltenv in content
+
+
+def test_cache_dest(mocked_opts, minion_opts):
+    """
+    Tests functionality for cache_dest
+    """
+    patched_opts = {x: y for x, y in minion_opts.items()}
```
Comment (s0undt3ch) [inline tests/pytests/unit/fileclient/test_fileclient_cache.py:312]:
```suggestion
    patched_opts = {x: y for x, y in minion_opts.items()}
```
Same as above

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

**Review by s0undt3ch — APPROVED** [review]

**Pull request merged**